Userland/Applications/Calculator/KeypadValue.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "KeypadValue.h"
+#include <AK/Math.h>
+#include <AK/String.h>
+
+KeypadValue::KeypadValue(i64 value, u8 decimal_places)
+    : m_value(value)
+    , m_decimal_places(decimal_places)
+{
+}
+
+KeypadValue::KeypadValue(i64 value)
+    : m_value(value)
+{
+}
+
+KeypadValue KeypadValue::operator+(KeypadValue const& rhs)
+{
+    return operate_on_keypadvalues<KeypadValue>(*this, rhs, [](KeypadValue const&, KeypadValue const& more_decimal_places, i64 less_decimal_places_equalized, i64 more_decimal_places_equalized, bool) -> KeypadValue {
+        return {
+            more_decimal_places_equalized + less_decimal_places_equalized,
+            more_decimal_places.m_decimal_places
+        };
+    });
+}
+
+KeypadValue KeypadValue::operator-(KeypadValue const& rhs)
+{
+    return *this + (-rhs); // NOTE: dirty little shorthand
+}
+
+KeypadValue KeypadValue::operator*(KeypadValue const& rhs)
+{
+    return operate_on_keypadvalues<KeypadValue>(*this, rhs, [](KeypadValue const& less_decimal_places, KeypadValue const& more_decimal_places, i64, i64, bool) -> KeypadValue {
+        return {
+            less_decimal_places.m_value * more_decimal_places.m_value,
+            (u8)(less_decimal_places.m_decimal_places + more_decimal_places.m_decimal_places)
+        };
+    });
+}
+
+KeypadValue KeypadValue::operator-(void) const
+{
+    return { -m_value, m_decimal_places };
+}
+
+bool KeypadValue::operator<(KeypadValue const& rhs)
+{
+    return operate_on_keypadvalues<bool>(*this, rhs, [](KeypadValue const&, KeypadValue const&, i64 less_decimal_places_equalized, i64 more_decimal_places_equalized, bool lhs_is_less) {
+        if (lhs_is_less)
+            return (less_decimal_places_equalized < more_decimal_places_equalized);
+        else
+            return (more_decimal_places_equalized < less_decimal_places_equalized);
+    });
+}
+
+bool KeypadValue::operator==(KeypadValue const& rhs)
+{
+    return operate_on_keypadvalues<bool>(*this, rhs, [](KeypadValue const&, KeypadValue const&, i64 less_decimal_places_equalized, i64 more_decimal_places_equalized, bool) {
+        return less_decimal_places_equalized == more_decimal_places_equalized;
+    });
+}
+
+KeypadValue::KeypadValue(double d)
+{
+    KeypadValue res = 0;
+    bool negative = false;
+    if (d < 0) {
+        negative = true;
+        d = -d;
+    }
+    i8 current_pow = 0;
+    while (AK::pow(10.0, (double)current_pow) <= d)
+        current_pow += 1;
+    current_pow -= 1;
+    while (d != 0) {
+        res.m_value *= 10;
+        res.m_value += (u64)(d / AK::pow(10.0, (double)current_pow)) % 10;
+        if (current_pow < 0)
+            res.m_decimal_places += 1;
+        current_pow -= 1;
+        if (res.m_decimal_places > 6)
+            break;
+    }
+    res = negative ? (-res) : res;
+    *this = res; //don't return ;)
```

Comment:
That's quite an odd comment, and an even more odd operation to put in a constructor